### PR TITLE
apps/speed.c: add -seconds and -bytes options

### PR DESCRIPTION
**Pull request by p-steuer**
Add speed tool options to run cipher, digest and rand benchmarks for a
single buffer size specified by -bytes over a time interval specified
by -seconds.

When sampling performance data its useful to run the benchmark for more seconds. Furthermore, one is usually interested in collecting samples for single buff size, not average over different sizes.

e.g.
```
openssl speed -seconds 10 -bytes 512 -evp aes-128-gcm
Doing aes-128-gcm for 10s on 512 size blocks: x aes-128-gcm's in 9.99s
[...]
The 'numbers' are in 1000s of bytes per second processed.
type            512 bytes
aes-128-gcm             xk
```

- [x] documentation is added or updated

### CODE REVIEW

File: apps/speed.c [hunk 1]
```diff
@@ -124,10 +124,19 @@
 #define MAX_ECDH_SIZE   256
 #define MISALIGN        64
 
+struct sec {
```
Comment (richsalz) [inline apps/speed.c]: I think make this a typedef, "typedef struct sec_st { .... } SEC" or similar.
And use time_t for the counters.

Comment (p-steuer) [reply apps/speed.c]: okay for the typedef.
problem with time_t is portability e.g. wrt conversion format specifier. The number of seconds was treated as int before i saw no reason to change this.

Comment (richsalz) [reply apps/speed.c]: okay that makes sense.  this looks good to me.

File: apps/speed.c [hunk 2]
```diff
@@ -1466,6 +1486,24 @@ int speed_main(int argc, char **argv)
             if (!opt_int(opt_arg(), &primes))
                 goto end;
             break;
+        case OPT_SECONDS:
+            seconds.sym = atoi(opt_arg());
+            seconds.rsa = atoi(opt_arg());
+            seconds.dsa = atoi(opt_arg());
+            seconds.ecdsa = atoi(opt_arg());
+            seconds.ecdh = atoi(opt_arg());
```
Comment (dot-asm) [inline apps/speed.c]: Even though it's as far from performance-critical path as only possible, I'd say that calling single call to `atoi(opt_arg())` would still be more appropriate. It's more readable that way, as you don't have to keep in mind how `opt_arg()` looks like. `seconds.sym = seconds.rsa = ... atoi(opt_arg())` would do :-)

File: doc/man1/speed.pod [hunk 3]
```diff
@@ -71,6 +71,14 @@ This can be used with a subsequent B<-rand> flag.
 Generate a B<num>-prime RSA key and use it to run the benchmarks. This option
 is only effective if RSA algorithm is specified to test.
 
+=item B<-seconds num>
+
+Run bechmarks for B<num> seconds.
+
+=item B<-bytes num>
+
+Run bechmarks on B<num>-byte buffers. Affects ciphers, digests and the CSPRNG.
```
Comment (dot-asm) [inline doc/man1/speed.pod]: This is not entirely true. as it doesn't affect multi-block benchmark :-)

Comment (p-steuer) [reply doc/man1/speed.pod]: oh, somehow overlooked that. fixed.

File: apps/speed.c (hunk 2 above)
Comment (p-steuer) [reply apps/speed.c]: agreed. fixed.

File: apps/speed.c [hunk 4]
```diff
@@ -1466,6 +1486,21 @@ int speed_main(int argc, char **argv)
...
             break;
+        case OPT_SECONDS:
+            seconds.sym = seconds.rsa = seconds.dsa = seconds.ecdsa
+                        = seconds.ecdh = atoi(opt_arg());
+            break;
+        case OPT_BYTES:
+            lengths_single = atoi(opt_arg());
+            if (lengths_single >= BUFSIZE) {
+                BIO_printf(bio_err,
+                           "%s: %d bytes exceeds max\n",
```
Comment (dot-asm) [inline apps/speed.c]: May I suggest to write even the max value? I mean if you expose this option to user, then it would be only appropriate to inform about the limit without having to refer to source code. I suppose one should print BUFSIZE-1. [Just in case, no, I don't know why it's defined with `+1`.]

On size note, multiblock_speed allocates own buffer and is free from this limitation.

Comment (dot-asm) [reply apps/speed.c]: Oh! While you are at it, just to spare a separate merge request, may I ask you to remove `(int)` casts in `app_malloc((int)BUFSIZE ...)`. [Just in case, no, I don't know why they were added back, but I'm sure it's redundant]. If you don't feel like it, it's fine, I'll submit separate request myself.

Comment (p-steuer) [reply apps/speed.c]: ok, got completely rid of BUFSIZE and SIZE_NUM limitations now. int casts were removed.

Comment (dot-asm) [reply apps/speed.c]: Cool! Thanks!

**Comment by dot-asm** [comment]
Just in case for reference. I used to do this, tailoring benchmarks for profiling, when needed, didn't think it was worth bothering users. Another test I occasionally tailor is buffer crossing page boundary ;-)

Ping @richsalz for re-approval.

**Comment by p-steuer** [comment]
Same for me, tired of doing the same changes for profiling over again. Anyway, default behaviour wasnt change, so no users bothered.

Some things i have in mind for speed tool:
```
                    if (decrypt)
                        EVP_DecryptInit_ex(loopargs[k].ctx, evp_cipher, NULL,
                                           key16, iv);
                    else
                        EVP_EncryptInit_ex(loopargs[k].ctx, evp_cipher, NULL,
                                           key16, iv);
```
Using 16 byte key here even in 256-bit case looks strange. Is there a rational behind this? I mean key16 is initialized to some specific value...

Moreover, id like to initialize cipher keys using EVP_CIPHER_CTX_rand_key, such that speed tool can handle engines that rely on specific format of key objects and/or non-standard keylengths.

**Comment by dot-asm** [comment]
> Using 16 byte key here even in 256-bit case looks strange. Is there a rational behind this? I mean key16 is initialized to some specific value...

key16 even in 256-bit case looks like bug. As for specific value, I don't know, never thought of that, random key would do just as well. Do you want to address it here, in this request?

**Comment by p-steuer** [comment]
Yes, i changed key32 to be the default.

Random keys will be a seperate PR. I ll also take a look at ctr mode ciphers which would need a reset in case 32 bit counter wrapping occurs (happens for larger -bytes and -seconds values).

**Comment by dot-asm** [comment]
> I ll also take a look at ctr mode ciphers which would need a reset in case 32 bit counter wrapping occurs (happens for larger -bytes and -seconds values).

Note that EVP_\*_ctr-s work with 128-bit counters. Even though low-level functions are called ctr32, high-level ones know how to and do handle it.

**Comment by p-steuer** [comment]
sorry, i mean ciphers with 32 bit ctrs like gcm or chacha. Im pretty sure i saw broken gcm profiling data due to overflow detection.

**Comment by richsalz** [comment]
re-approved.  As for the page-boundary, is it more than -misaligned you use?

**Comment by dot-asm** [comment]
> As for the page-boundary, is it more than -misaligned you use?

Yes. -misalign is just few bytes misalign, up to cache line, if one wants to catch cache line boundaries in short inputs. Note that big buffer is guaranteed to cross page boundary, several in fact, so that aiming for page boundaries is a small-blocks thing...

**Comment by dot-asm** [comment]
Merged. Thanks.
